### PR TITLE
feat(module:date-picker): support parse input value

### LINKED ISSUE #2492: Problem to set date manually datepicker when chage nzFormat
## Version
1.8.0
## Environment
Windows
## Reproduction link
[https://ng-zorro-antd-start-cekrxg.stackblitz.io](https://ng-zorro-antd-start-cekrxg.stackblitz.io)
## Steps to reproduce
Set dateformat to yyyy-MM-dd in the nz-datepicket, after set date manually in the input box datepicker a date for example 2018-01-01
## What is expected?
Ever
## What is actually happening?
Problem to set manually date when dateFormat is changed. For example, if you set nzFormat to yyyy/MM/dd set date manually work fine, the datepicker set a date like 2018/01/01 then hide, this is ok. But when nzFormat is set to yyyy-MM-dd, set date manually not work, the datepicker is maintain visible and if you does click outside the calendar the date is not set

### LINKED ISSUE #3976: DatePicker doesn't recognize manual date input when using nzFormat
### Reproduction link
[https://stackblitz.com/edit/angular-a8d35n](https://stackblitz.com/edit/angular-a8d35n)

### Steps to reproduce
Add DatePicker component and set nzFormat="dd.MM.yyyy".
```javascript
<nz-date-picker [(ngModel)]="date" nzFormat="dd.MM.yyyy"></nz-date-picker>
```
Manual input works only if date and month are same. For example, if we type 04.04.1995 it works, but 04.05.1995 is not working.

### What is expected?
After writing string date, DatePicker changes date in the model.

### What is actually happening?
Nothing happens

| Environment | Info |
|---|---|
| ng-zorro-antd | 8.1.2 |
| Browser | Chrome 76 |

---
Default format working fine

### LINKED ISSUE #4028: Enter key selects 2001 on nz-date-picker
### Reproduction link
[https://stackblitz.com/edit/ng-zorro-antd-start-ljdvgm](https://stackblitz.com/edit/ng-zorro-antd-start-ljdvgm)

### Steps to reproduce
1. Click date-picker input box
2. Pick date
3. Click 'ok'
4. Click date-picker input box (again)
5. Press 'enter' key. Notice year changes to 2001.

### What is expected?
Enter key will close date-picker.

### What is actually happening?
Enter changes year 2001.

| Environment | Info |
|---|---|
| ng-zorro-antd | 8.2.0 |
| Browser | Version 76.0.3809.100 (Official Build) (64-bit) |

### LINKED ISSUE #4101: 日期选择框最上方的input框修改日期无效
### Reproduction link
[https://ng-zorro-antd-start.stackblitz.io](https://ng-zorro-antd-start.stackblitz.io)

### Steps to reproduce
点击日期选择框，然后手动修改上方input里面的日期，点击确认时间不变

### What is expected?
时间变成最新修改的时间

### What is actually happening?
时间还是自己选中的时间，不是修改以后的时间

| Environment | Info |
|---|---|
| ng-zorro-antd | 8.2.1 |
| Browser |  76.0.3809.132 |

### PR DESCRIPTION
**Pull request by wenqi73**
Close #4028 
Close #3976
Close #2492
Close #4101

## 新功能
date-fns 1.x 不支持日期的反格式化，相关 issue 有：#4028 #3976 #2492 #4101。
date-fnx 2.x 已支持[该功能](https://date-fns.org/v2.9.0/docs/parse)。

date-fns 1.x does not support date parsing. Related issues are: #4028 #3976 #2492 #4101. date-fnx 2.x already supports [this feature] (https://date-fns.org/v2.9.0/docs/parse)

## 如何实现
由于 angular 的[日期格式](https://angular.io/api/common/DatePipe)与 date-fns 的[日期格式](https://date-fns.org/v2.9.0/docs/parse)并不兼容。因此反格式化功能只有在用户引入了 date-fns 才能使用。

Since [angular's date format](https://angular.io/api/common/DatePipe) and [date-fns' date format](https://date-fns.org/v2.9.0/docs/parse) are not compatible. Therefore, the parsing feature could only be supported when the user provides date-fns.

## Breaking Change
我们推荐用户使用 date-fns 2.x 的新的 format 格式。同时也提供配置项将 1.x format 转为 2.x format，新旧 format 的对比参照[这里](https://github.com/date-fns/date-fns/blob/master/CHANGELOG.md#200---2019-08-20)。

We recommend that users use the new format of date-fns 2.x, and we also provide configuration items to convert 1.x format to 2.x format. For a comparison of the old and new formats, refer to [here](https://github.com/date-fns/date-fns/blob/master/CHANGELOG.md#200---2019-08-20).

```ts
@NgModule({
  ...
  imports     : [ NgZorroAntdModule ],
  providers: [
    { provide: NZ_DATE_LOCALE, useValue: enUS },
    // Set `dateFnsFormatConvert ` to convert format from v1 to v2
    { provide: NZ_DATE_FNS_COMPATIBLE, useValue: true }
  ]
})
export class AppModule { }
```

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Application (the showcase website) / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?
```
[x] Yes
[ ] No
```

### CODE REVIEW

File: components/i18n/nz-i18n.interface.ts [hunk 1]
```diff
@@ -119,4 +120,4 @@ export interface NzI18nInterface {
 }
 
 // tslint:disable-next-line:no-any
-export type DateLocale = any; // TODO: Implement this type definition when date-fns is stable
+export type DateLocale = Locale; // TODO: Implement this type definition when date-fns is stable
```
Comment (hsuanxyz) [inline components/i18n/nz-i18n.interface.ts]:
```suggestion
export type DateLocale = Locale;
```

File: docs/i18n.zh-CN.md [hunk 2]
```diff
@@ -42,6 +42,15 @@ export class AppModule { }
 }
 ```
 
+### NZ_DATE_FORMAT_CONVERT
+
+我们已将 `date-fns` 升级至 2.x，**我们推荐用户使用 `date-fns` 2.x 的日期格式**。同时也提供 `NZ_DATE_FORMAT_CONVERT`，当设置为 `true` 时，`ng-zorro-antd` 会把 1.x 的格式转为 2.x，新旧格式的对比参照[这里](https://github.com/date-fns/date-fns/blob/master/CHANGELOG.md#200---2019-08-20)。
```
Comment (hsuanxyz) [inline docs/i18n.zh-CN.md]: 这里给警告，下版本移除 1.x 支持

File: components/i18n/date-config.ts [hunk 3]
```diff
@@ -7,18 +7,40 @@
...
+ * Compatible translate the moment-like format pattern to angular's pattern
+ * Why? For now, we need to support the existing language formats in AntD, and AntD uses the default temporal syntax.
+ *
+ * TODO: compare and complete all format patterns
+ * Each format docs as below:
+ * @link https://momentjs.com/docs/#/displaying/format/
+ * @link https://angular.io/api/common/DatePipe#description
+ * @param format input format pattern
+ */
+export function transCompatFormat(format: string): string {
```
Comment (hsuanxyz) [inline components/i18n/date-config.ts]: 移到 date-pick 中

File: components/i18n/convert-tokens.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+// tslint:disable-next-line:prefer-template
+const tokensRegExp = new RegExp('(\\[[^\\[]*\\])|(\\\\)?' + '(' + v1tokens.join('|') + '|.)', 'g');
+
+interface TokensBuffer {
+  formatBuffer: string[];
+  textBuffer: string[];
+}
+
+export function convertTokens(format: string): string {
```
[the remark this answers is not in the document]
Comment (wenqi73) [reply components/i18n/convert-tokens.ts]: i18n/date-helper.service 要用到，移了 build 报错

File: components/date-picker/lib/date-header.component.ts [hunk 5]
```diff
@@ -25,16 +26,12 @@ export class DateHeaderComponent extends AbstractPanelHeader {
   }
 
   getSelectors(): PanelSelector[] {
```
Comment (hsuanxyz) [inline components/date-picker/lib/date-header.component.ts]: 检查下 i18n 改变时，`getSelectors` 有没有重新取值

File: components/i18n/convert-tokens.ts [hunk 6]
```diff
@@ -72,6 +73,7 @@ interface TokensBuffer {
 }
 
 export function convertTokens(format: string): string {
+  warnDeprecation(`'NZ_DATE_FNS_COMPATIBLE' is deprecated and will be removed in next minor version. Please update to date-fns v2 format.`);
```
Comment (hsuanxyz) [inline components/i18n/convert-tokens.ts]:
```suggestion
  warnDeprecation(`'NZ_DATE_FNS_COMPATIBLE' Will be removed in 10.0.0, please update to date-fns v2 format`);
```

**Comment by valerio-lucantonio** [comment]
Hi, with which new version will this fix be delivered? thanks

**Comment by vthinkxie** [comment]
9.0.0-beta.1

**Pull request merged**